API/models/rbac.go

```diff
@@ -0,0 +1,122 @@
+package models
+
+import (
+	u "p3/utils"
+	"strings"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// Roles
+const (
+	Manager = "manager"
+	User    = "user"
+	Viewer  = "viewer"
+)
+
+// Actions
+const (
+	READ = iota
+	WRITE
+	READONLYNAME
+)
+
+const ROOT_DOMAIN = "*"
+
+func CheckDomainExists(domain string) bool {
+	if domain == ROOT_DOMAIN {
+		return true
+	}
+	x, e := GetEntity(bson.M{"hierarchyName": domain}, "domain", u.RequestFilters{}, nil)
+	return e == "" && x != nil
+}
+
+func GetRequestFilterByDomain(userRoles map[string]string) (bson.M, bool) {
+	filter := bson.M{}
+	if userRoles[ROOT_DOMAIN] == Manager || userRoles[ROOT_DOMAIN] == User {
+		return filter, true
+	}
+	domainPattern := ""
+	for domain, role := range userRoles {
+		switch role {
+		case User:
+		case Manager:
+			if domainPattern == "" {
+				domainPattern = domain
+			} else {
+				domainPattern = domainPattern + "|" + domain
+			}
+		}
+	}
+	if domainPattern == "" {
+		return filter, false
+	} else {
+		filter["domain"] = primitive.Regex{Pattern: domainPattern, Options: ""}
+		return filter, true
+	}
+}
+
+func CheckUserPermissions(userRoles map[string]string, objEntity, requestType int, objDomain string) (bool, int) {
```

Comment:
I do not understand why the function has to return the permission, can it be different from the one requested ? I assume maybe it is to return "readonlyname" when "read" is requested, but I think it is less confusing to simply return false in that case.

Reply:
I believe it is confusing, the boolean could be misinterepreted as whether or not the user has the permission requestType, which is not the case because right now it will return true if the user only has READONLYNAME while requestType is READ.
I think it is better to either

-  remove requestType, and only return the best permission, possibly None. This works because write implies read and read implies readonlyname, so the caller gets all the information he needs.

or

-   keep requestType and return only the boolean with the meaning "the user has the permission requestType". In that case the function can be called a second time if we want to check for readonly, the performance hit is small.